lib/AST/Decl.cpp

```diff
@@ -3027,6 +3027,9 @@ bool ProtocolDecl::existentialConformsToSelfSlow() {
 static SelfReferenceKind
 findProtocolSelfReferences(const ProtocolDecl *proto, Type type,
                            bool skipAssocTypes) {
+  // InOut types are invariant.  Look through them.
+  type = type->getInOutObjectType();
```

Comment:
I don't believe this is correct. A function type taking an inout Self is not contravariant, it is invariant.

Reply:
Contravariant?

Reply:
findProtocolSelfReferences() with `(Self) -> ()` should return a `SelfReferenceKind` with `parameter=true` and all other fields false. But if you pass in `(inout Self) -> ()` it should set `other=true`. I think with your change here, it just ignores the inout, which is wrong.

Reply:
Here is a test case:

```
Slavas-MacBook-Pro:swift slava$ cat io.swift 
protocol P {
  func f(_: inout Self)
}

class C : P {
  func f(_: inout C) {}
}
Slavas-MacBook-Pro:swift slava$ xcrun swiftc io.swift
io.swift:6:8: error: protocol 'P' requirement 'f' cannot be satisfied by a non-final class ('C') because it uses 'Self' in a non-parameter, non-result type position
  func f(_: inout C) {}
       ^

```

Reply:
Ah.  I'll revert this change then.